server/service/integration_core_test.go

```diff
@@ -3601,6 +3600,110 @@ func (s *integrationTestSuite) TestDeviceAuthenticatedEndpoints() {
 	res.Body.Close()
 }
 
+func (s *integrationTestSuite) TestModifyUser() {
+	t := s.T()
+
+	// create a new user
+	var createResp createUserResponse
+	userRawPwd := "passw0rd!"
+	params := fleet.UserPayload{
+		Name:                     ptr.String("moduser"),
+		Email:                    ptr.String("moduser@example.com"),
+		Password:                 ptr.String(userRawPwd),
+		GlobalRole:               ptr.String(fleet.RoleObserver),
+		AdminForcedPasswordReset: ptr.Bool(false),
+	}
+	s.DoJSON("POST", "/api/v1/fleet/users/admin", params, http.StatusOK, &createResp)
+	require.NotZero(t, createResp.User.ID)
+	u := *createResp.User
+
+	s.token = s.getTestToken(u.Email, userRawPwd)
+	require.NotEmpty(t, s.token)
+	defer func() { s.token = s.getTestAdminToken() }()
+
+	// as the user: modify email without providing current password
+	var modResp modifyUserResponse
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		Email: ptr.String("moduser2@example.com"),
+	}, http.StatusUnprocessableEntity, &modResp)
+
+	// as the user: modify email with invalid password
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		Email:    ptr.String("moduser2@example.com"),
+		Password: ptr.String("nosuchpwd"),
+	}, http.StatusForbidden, &modResp)
+
+	// as the user: modify email with current password
+	newEmail := "moduser2@example.com"
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		Email:    ptr.String(newEmail),
+		Password: ptr.String(userRawPwd),
+	}, http.StatusOK, &modResp)
+	require.Equal(t, u.ID, modResp.User.ID)
+	require.Equal(t, u.Email, modResp.User.Email) // new email is pending confirmation, not changed immediately
+
+	// as the user: set new password without providing current one
+	newRawPwd := userRawPwd + "2"
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		NewPassword: ptr.String(newRawPwd),
+	}, http.StatusUnprocessableEntity, &modResp)
+
+	// as the user: set new password with an invalid current password
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		NewPassword: ptr.String(newRawPwd),
+		Password:    ptr.String("nosuchpwd"),
+	}, http.StatusForbidden, &modResp)
+
+	// as the user: set new password and change name, with a valid current password
+	modResp = modifyUserResponse{}
+	s.DoJSON("PATCH", fmt.Sprintf("/api/v1/fleet/users/%d", u.ID), fleet.UserPayload{
+		NewPassword: ptr.String(newRawPwd),
+		Password:    ptr.String(userRawPwd),
+		Name:        ptr.String("moduser2"),
+	}, http.StatusOK, &modResp)
+	require.Equal(t, u.ID, modResp.User.ID)
+	require.Equal(t, "moduser2", modResp.User.Name)
```

Comment:
As can be seen in the tests above, I went with a behaviour similar to the change_password endpoint for consistency, regarding this comment that I left on the ticket: https://github.com/fleetdm/fleet/issues/3059#issuecomment-1063405562.